node/network/src/protocol.rs

```diff
@@ -1,24 +1,108 @@
+use configs::NetworkConfig;
 use primitives::beacon::SignedBeaconBlock;
+use primitives::chain::ChainPayload;
 use primitives::chain::SignedShardBlock;
+use primitives::network::PeerInfo;
+use primitives::serialize::{Decode, Encode};
+use primitives::types::{AccountId, Gossip, PeerId};
+
+use crate::message::{CoupledBlock, Message, RequestId};
+use crate::peer::PeerMessage;
+use crate::peer_manager::PeerManager;
+
+/// Protocol responsible for actual message processing from network and sending messages.
+struct Protocol {
+    client: Arc<Client>,
+    peer_manager: Arc<PeerManager>,
+    inc_gossip_tx: Sender<Gossip<ChainPayload>>,
+    inc_block_tx: Sender<CoupledBlock>,
+}
+
+impl Protocol {
+    fn receive_message(&self, peer_id: PeerId, data: Vec<u8>) {
+        match Decode::decode(&data) {
+            Ok(m) => match m {
+                Message::Gossip(gossip) => {
+                    forward_msg(self.inc_gossip_tx.clone(), *gossip)
+                },
+                Message::BlockAnnounce(block) => {
+                    let unboxed = *block;
+                    forward_msg(self.inc_block_tx.clone(), (unboxed.0, unboxed.1));
+                },
+                Message::BlockFetchRequest(request_id, hashes) => {
+                    match self.client.fetch_blocks(&hashes) {
+                        Ok(blocks) => self.send_block_response(&peer_id, request_id, blocks),
+                        Err(_) => {
+                            self.peer_manager.suspect_malicious(&peer_id);
+                            warn!("Failed to fetch blocks for {:?} from {}. Possible grinding attack.", hashes, peer_id);
+                        }
+                    }
+
+                },
+                Message::PayloadRequest(request_id, transaction_hashes, receipt_hashes) => {
+                    match self.client.fetch_payload(transaction_hashes, receipt_hashes) {
+                        Ok(payload) => self.send_payload_response(&peer_id, request_id, payload),
+                        Err(_) => {
+                            self.peer_manager.suspect_malicious(&peer_id);
+                            warn!("Failed to fetch payload for {:?}, {:?} from {}. Possible grinding attack.", transaction_hashes, receipt_hashes, peer_id);
+                        }
+                    }
+                }
+                _ => (),
+            },
+            Err(e) => warn!(target: "network", "{}", e),
+        };
+    }
+
+    fn send_gossip(&self, g: Gossip<ChainPayload>) {
+        let auth_map = self.client.get_recent_uid_to_authority_map();
+        let out_channel = auth_map
+            .get(&g.receiver_uid)
+            .map(|auth| auth.account_id.clone())
+            .and_then(|account_id| self.peer_manager.get_account_channel(account_id));
+        if let Some(ch) = out_channel {
+            let data = Encode::encode(&Message::Gossip(Box::new(g))).unwrap();
+            forward_msg(ch, PeerMessage::Message(data));
+        }
```

Comment:
I would add `else` and print `warn` saying the channel was not found for the given uid, it proved to be very useful when debugging.